### PR TITLE
feat(model): Support flash-attention (non-streaming version)

### PR DESCRIPTION
**Pull request by xingchensong**
https://github.com/Dao-AILab/flash-attention

## TODO
- [ ] unit test
- [ ] benchmark

### CODE REVIEW

**Comment by xingchensong** [comment]
flash_attn 安装，cuda版本需要 >= 11.6，在这里纪录下安装方式

```sh
cuda_version=11.6
export PATH=/usr/local/cuda-${cuda_version}/bin${PATH:+:${PATH}}
export CUDA_HOME=/usr/local/cuda-${cuda_version}
export CFLAGS="-I$CUDA_HOME/include $CFLAGS"
export LD_LIBRARY_PATH=/usr/local/cuda-${cuda_version}/lib64:$LD_LIBRARY_PATH                                                                                                                                                        
export CUDA_PATH=$CUDA_HOME
pip install flash_attn --no-build-isolation
```

**Comment by xingchensong** [comment]
why `use_reentrant=False` for gradient checkpointing?

1. It is recommended by torch, (even though currently the default is `True`) https://pytorch.org/docs/stable/checkpoint.html
2. there seems a bug when enable `grad_ckpt + use_reentrant=True` and `flash_selfattn`

**Comment by xingchensong** [comment]
Please note that flash_attn only supports `fp16` or `bf16`, we cannot use flash_attn under `fp32` training.

**Comment by xingchensong** [comment]
Because FlashAttn can only run on GPU, considering that the CI/CD machines on GitHub may not have GPUs, I will place the unit test code in this comment. The unit test results show that the numerical precision of forward-pass output values is `1e-2`. As for backward-pass,  the numerical precision of grad is `4.2`, this is weird.

```py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright [2023-12-05] <sxc19@mails.tsinghua.edu.cn, Xingchen Song>

import torch
import numpy as np

from wenet.utils.mask import make_non_pad_mask
from wenet.transformer.attention import MultiHeadedAttention, FlashMultiHeadedAttention


def test_flash_attn():
    torch.manual_seed(777)
    # Init model
    mha = MultiHeadedAttention(n_head=16, n_feat=256,
                               dropout_rate=0.0, key_bias=True)
    flash_mha = FlashMultiHeadedAttention(n_head=16, n_feat=256,
                                          dropout_rate=0.0, key_bias=True)
    flash_mha.load_state_dict(mha.state_dict(), strict=True)
    mha.cuda().train().half()
    flash_mha.cuda().train().half()
    # Forward
    q = torch.randn(3, 10, 256).cuda().half() + 10.0
    k = torch.randn(3, 10, 256).cuda().half() + 11.0
    v = torch.randn(3, 10, 256).cuda().half() + 12.0
    q[0, 1:, :], k[0, 1:, :], v[0, 1:, :] = 0.0, 0.0, 0.0
    q[1, 5:, :], k[1, 5:, :], v[1, 5:, :] = 0.0, 0.0, 0.0
    mask = make_non_pad_mask(torch.tensor([1, 5, 10])).unsqueeze(1).cuda()  # (3, 1, 10)
    out = mha(q, k, v, mask)[0]
    flash_out = flash_mha(q, k, v, mask)[0]
    indices, _, _ = flash_mha.unpad(mask)
    out = out.reshape(30, 256)[indices]
    flash_out = flash_out.reshape(30, 256)[indices]
    np.testing.assert_allclose(out.detach().cpu().numpy(),
                               flash_out.detach().cpu().numpy(),
                               rtol=1e-7, atol=1e-2)
    # backward
    mha.zero_grad()
    out.sum().backward()
    flash_mha.zero_grad()
    flash_out.sum().backward()
    for (name1, param1), (name2, param2) in zip(
            mha.named_parameters(), flash_mha.named_parameters()):
        assert name1 == name2
        if name1 in ["linear_q.weight", "linear_k.weight", "linear_v.weight",
                     "linear_q.bias", "linear_k.bias", "linear_v.bias"]:
            print("Check {}".format(name1))
            np.testing.assert_allclose(
                param1.grad.detach().cpu().numpy(),
                param2.grad.detach().cpu().numpy(),
                rtol=1e-7, atol=4.2)
            print("Pass {}".format(name1))
```

**Comment by xingchensong** [comment]
IMPL ref:
1. https://github.com/espnet/espnet/pull/5537/files#diff-812fc96e91a83bdb1f453b203efef68bef6c3113377ebf0a5cf0828fd6e0815bR129-R183
2. https://github.com/huggingface/transformers/pull/27203/files#diff-6ec07359a4b7f7c8c986c86549f01867ea071bef35cac64e0bd05302fef2e615R472-R681

**Comment by xingchensong** [comment]
Flash attention is now natively supported starting from Torch 2.1 and onwards. Consequently, this pull request will be closed, and the implementation of flash attention will proceed in  https://github.com/wenet-e2e/wenet/pull/2191
